crates/sui-indexer/src/store/pg_indexer_analytical_store.rs

```diff
@@ -0,0 +1,374 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use async_trait::async_trait;
+use core::result::Result::Ok;
+use std::time::Duration;
+
+use diesel::dsl::{count, max};
+use diesel::upsert::excluded;
+use diesel::ExpressionMethods;
+use diesel::{QueryDsl, RunQueryDsl};
+
+use crate::types_v2::IndexerResult;
+use crate::PgConnectionPool;
+
+use crate::schema_v2::{
+    active_addresses, address_metrics, addresses, checkpoints, move_call_metrics, move_calls,
+    network_metrics, transactions, tx_count_metrics, tx_indices,
+};
+
+use crate::errors::{Context, IndexerError};
+use crate::models_v2::address_metrics::{StoredActiveAddress, StoredAddress, StoredAddressMetrics};
+use crate::models_v2::checkpoints::StoredCheckpoint;
+use crate::models_v2::move_call_metrics::{
+    build_move_call_metric_query, QueriedMoveCallMetrics, QueriedMoveMetrics, StoredMoveCall,
+    StoredMoveCallMetrics,
+};
+use crate::models_v2::network_metrics::{RowCountEstimation, StoredNetworkMetrics};
+use crate::models_v2::transactions::StoredTransaction;
+use crate::models_v2::tx_count_metrics::StoredTxCountMetrics;
+use crate::models_v2::tx_indices::StoredTxIndex;
+
+use super::IndexerAnalyticalStore;
+use crate::store::diesel_macro::{read_only_blocking, transactional_blocking_with_retry};
+
+const PG_COMMIT_CHUNK_SIZE: usize = 1000;
+
+#[derive(Clone)]
+pub struct PgIndexerAnalyticalStore {
+    blocking_cp: PgConnectionPool,
+}
+
+impl PgIndexerAnalyticalStore {
+    pub fn new(blocking_cp: PgConnectionPool) -> Self {
+        Self { blocking_cp }
+    }
+}
+
+#[async_trait]
+impl IndexerAnalyticalStore for PgIndexerAnalyticalStore {
+    async fn get_latest_stored_checkpoint(&self) -> IndexerResult<StoredCheckpoint> {
+        let latest_cp = read_only_blocking!(&self.blocking_cp, |conn| {
+            checkpoints::dsl::checkpoints
+                .order(checkpoints::sequence_number.desc())
+                .first::<StoredCheckpoint>(conn)
+        })
+        .context("Failed reading latest checkpoint from PostgresDB")?;
+        Ok(latest_cp)
+    }
+
+    async fn get_checkpoints_in_range(
+        &self,
+        start_checkpoint: i64,
+        end_checkpoint: i64,
+    ) -> IndexerResult<Vec<StoredCheckpoint>> {
+        let cps = read_only_blocking!(&self.blocking_cp, |conn| {
+            checkpoints::dsl::checkpoints
+                .filter(checkpoints::sequence_number.ge(start_checkpoint))
+                .filter(checkpoints::sequence_number.lt(end_checkpoint))
+                .order(checkpoints::sequence_number.asc())
+                .load::<StoredCheckpoint>(conn)
+        })
+        .context("Failed reading checkpoints from PostgresDB")?;
+        Ok(cps)
+    }
+
+    async fn get_transactions_in_checkpoint_range(
+        &self,
+        start_checkpoint: i64,
+        end_checkpoint: i64,
+    ) -> IndexerResult<Vec<StoredTransaction>> {
+        let tx_batch = read_only_blocking!(&self.blocking_cp, |conn| {
```

Comment:
Would you be able to change from using these macros to using a the function based approach that we're using in indexer_reader.rs? Its a bit more ergonomic and easier to read.

Reply:
I was looking at pg_indexer_store_v2, we had macros there as well for both read-write and read-only.
yeah I like the function based way better too, happy to refactor them together in a following PR.

Reply:
sounds good, we can do that as a followup